pkg/function/restore_csi_snapshot.go

```diff
@@ -0,0 +1,197 @@
+		return nil, err
+	}
+	if err := Arg(args, RestoreCSISnapshotNamespaceArg, &restoreArgs.Namespace); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, RestoreCSISnapshotStorageClassArg, &restoreArgs.StorageClass); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, RestoreCSISnapshotRestoreSizeArg, &restoreSize); err != nil {
+		return nil, err
+	}
+	if err := OptArg(args, RestoreCSISnapshotAccessModesArg, &restoreArgs.AccessModes, []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce}); err != nil {
+		return nil, err
+	}
+	if err := validateVolumeAccessModesArg(restoreArgs); err != nil {
+		return nil, err
+	}
+	if err := OptArg(args, RestoreCSISnapshotVolumeModeArg, &restoreArgs.VolumeMode, v1.PersistentVolumeFilesystem); err != nil {
+		return nil, err
+	}
+	if err := validateVolumeModeArg(restoreArgs); err != nil {
+		return nil, err
+	}
+	if err := OptArg(args, RestoreCSISnapshotLabelsArg, &restoreArgs.Labels, nil); err != nil {
+		return nil, err
+	}
+	size, err := resource.ParseQuantity(restoreSize)
+	if err != nil {
+		return nil, err
+	}
+	restoreArgs.RestoreSize = &size
+
+	if err := restoreCSISnapshot(ctx, restoreArgs); err != nil {
+		return nil, err
+	}
+	return nil, nil
+}
+
+func (*restoreCSISnapshotFunc) RequiredArgs() []string {
+	return []string{
+		RestoreCSISnapshotNameArg,
+		RestoreCSISnapshotPVCNameArg,
+		RestoreCSISnapshotNamespaceArg,
+		RestoreCSISnapshotStorageClassArg,
+		RestoreCSISnapshotRestoreSizeArg,
+	}
+}
+
+func restoreCSISnapshot(ctx context.Context, args restoreCSISnapshotArgs) error {
+	kubeCli, err := kube.NewClient()
+	if err != nil {
+		return err
+	}
+	pvc := newPVCManifest(args)
+	if _, err = kubeCli.CoreV1().PersistentVolumeClaims(args.Namespace).Create(ctx, pvc, metav1.CreateOptions{}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func newPVCManifest(args restoreCSISnapshotArgs) *v1.PersistentVolumeClaim {
+	snapshotAPIGroup := SnapshotAPIGroup
+	pvc := &v1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      args.PVC,
+			Namespace: args.Namespace,
+		},
+		Spec: v1.PersistentVolumeClaimSpec{
+			AccessModes: args.AccessModes,
+			VolumeMode:  &args.VolumeMode,
+			DataSource: &v1.TypedLocalObjectReference{
+				APIGroup: &snapshotAPIGroup,
+				Kind:     "VolumeSnapshot",
+				Name:     args.Name,
+			},
+			StorageClassName: &args.StorageClass,
+			Resources: v1.ResourceRequirements{
+				Requests: v1.ResourceList{
+					v1.ResourceStorage: *args.RestoreSize,
+				},
+			},
+		},
+	}
+	if args.Labels != nil {
+		pvc.ObjectMeta.Labels = args.Labels
+	}
+	return pvc
+}
+
+func validateVolumeModeArg(args restoreCSISnapshotArgs) error {
+	for _, validVolumeMode := range []v1.PersistentVolumeMode{
+		v1.PersistentVolumeFilesystem,
+		v1.PersistentVolumeBlock,
+	} {
+		if args.VolumeMode != validVolumeMode {
+			return errors.New("Given volumeMode" + string(args.VolumeMode) + " is invalid")
+		}
+	}
+	return nil
+}
```

Comment:
this seems a bit wrong to me, if users specified the volumeMode to be `v1.PersistentVolumeBlock`, wouldn't the for loop return in first iteration saying given volume mode is invalid.